### ISSUE TITLE
Build on save, but not run tests?

### ISSUE THREAD

**Issue by TheChuckster** [Microsoft/vscode-go#451]
Is this possible? My unit tests are stateful and not concurrent (yeah, yeah...), and so I can't just have them run in the background arbitrarily.

**Comment by lukehoban** [comment]
This should be what happens already (build, but not run). What makes you think tests are being run?

**Comment by TheChuckster** [comment]
Two things:

1. My tests hit my web service -- I see my web service getting hit with the tests whenever VSCode is running and I save my main.go source file (and obviously, I'm not running "go test" or anything like that on the command line).

2. From your repository's README.md markdown (line 19) mentions that vs-code does a "build-on-save (using go build and go test)"

**Comment by DasJott** [comment]
One could think, that tests are run on save, bcause of your doc in the settings:

```
// Go configuration

	// Run 'go build'/'go test' on save.
	"go.buildOnSave": true,

  ...
```

**Comment by ramya-rao-a** [comment]
Looking at the code, buildOnSave runs `go build` on the current package if current file is not a test file i.e filename ending with "_test.go"
If file is a test file then `go test` is run on the current package instead of `go build`.

Note: When `go test` is run this way, the output of the test does not reach the output channel.
@DasJott maybe this is why you say that the tests are not run?

We can have a separate flag `go.testOnSave` that would run the `go test` on the current package and separate testing feature from `go.buildOnSave`

**Comment by ramya-rao-a** [comment]
@lukehoban Any thoughts on the reason behind the original design to run `go test` on the current package if current file is a test file instead of just `go build` when `go.buildOnSave` is true?

**Comment by ironcladlou** [comment]
Prior to this issue I hadn't noticed that `buildOnSave` was described as having behavior for tests. That particular functionality predates my involvement in the extension.

That said, separating `testOnSave` from `buildOnSave` makes good sense to me.

**Comment by lukehoban** [comment]
To build the tests and get compiler errors, you have to run `go test -c`, which compiles the tests but does not run them.  This is what the extension does.  You should not see any tests being executed by `buildOnSave`.  It's not clear to me that we really are seeing tests get run - it would be great to make sure we have a repro for this.

I think the fix may be as simple as changing the documentation from: `"Run 'go build'/'go test' on save."` to `"Run 'go build'/'go test -c' on save."` or `"Run 'go build'' on save."` (leaving this subtle detail that `go build` doesn't build tests as unmentioned).

**Comment by ramya-rao-a** [comment]
@lukehoban my bad and thanks for clarifying, I didn't know that `go build`does not build test files and I totally missed the `-c` flag in the `go test`.

Updating the documentation makes sense.

@TheChuckster This still doesn't solve your original query. Did you figure that out or you still feel that your tests are running on save?

**Comment by DasJott** [comment]
@ramya-rao-a That could indeed be the reason. I might just have the impression that the tests are not run.
I like the idea of having seperate options. Additionally I would like to be able to put options for executing all tests, like `go test ./...` does, because I would like all tests to be run on save. Should be no problem, as there is no blocking.

**Comment by ramya-rao-a** [comment]
Created another issue to track the test on save feature.

The comment on `go.buildOnSave`  has been updated.

Closing this issue, as it is clear now that tests are NOT run when `go.buildOnSave` is set to true

**Comment by ramya-rao-a** [comment]
The `testOnSave` feature is now available in the latest update (0.6.54) to the Go extension.
